### DISCUSSION TITLE
For a fallible void function, do you use Option<Exception> or Either<Exception, Unit>?

### DISCUSSION THREAD

**Discussion by petejodo** [arrow-kt/arrow#2907] [General]
I couldn't find anything on this topic. Title covers it pretty well, basically I'm looking to determine what the arrow devs and community typically do for a function like

```kotlin
fun foo(): Option<Exception> {}

// vs

fun foo(): Either<Exception, Unit> {}
```

I ask this because different communities do different things. For example, the Rust community uses `Result<Err, ()>` (`Either`) which I personally prefer but the naming of "Result" is better suited towards that than `Either` and then in Haskell they use `Maybe` (`Option`). What does the kotlin/arrow community prefer?

**Comment by nomisRev** [comment]
Hey @petejodo,
I've never seen `Option<Exception>` in Haskell 🤔 Typically `Option` is used for modelling _abscene_ of a value, and most often _nullable types_ are preferred for this.

In Kotlin with Arrow I would suggest using `Either`, since operators like `Either.catch` automatically take care of `CancellationException` and _fatal_ exceptions. You could introduce `typealias Result<E, A> = Either<E, A>` if you prefer that name.

Alternatively you can use `kotlin.Result` and Arrow has support for `kotlin.Result` in it's APIs exposing `result { }` DSL similarly to `either { }` DSL. Take care though since `kotlin.Result`'s `runCatching` will swallow `CancellationException` and _fatal_ exceptions so you need to manually deal with them.

**Reply by petejodo** [reply]
hmm yeah... take my note about Haskell and put it in a `Maybe` 😏 but ok yeah that's how I was thinking about it but wasn't entirely sure. I also asked this before I noticed the `void()` method on result and so I was doing an ugly `.map { }`. I found that to be a clue into the preference of `Either` in such cases

Also, I did not know that about arrow's support around kotlin's result, thanks for the tip!

**Comment by kyay10** [comment]
I'd personally go with `Either<Exception, Unit>`, `Result<Unit>`, or:
```kotlin
context(_: Raise<Exception>) fun foo {}
```
Which I believe to be the most idiomatic.
